bindings/matrix-sdk-ffi/src/authentication_service.rs

```diff
@@ -0,0 +1,111 @@
+pub enum AuthenticationError {
+    #[error("A successful call to use_server must be made first.")]
+    ClientMissing,
+    #[error("An error occurred: {message}")]
+    Generic { message: String },
+}
+
+impl From<anyhow::Error> for AuthenticationError {
+    fn from(e: anyhow::Error) -> AuthenticationError {
+        AuthenticationError::Generic { message: e.to_string() }
+    }
+}
+
+impl AuthenticationService {
+    /// Creates a new service to authenticate a user with.
+    pub fn new(base_path: String) -> Self {
+        AuthenticationService {
+            base_path,
+            client_container: RwLock::new(ClientContainer { client: None }),
+        }
+    }
+
+    /// The currently configured homeserver.
+    pub fn homeserver(&self) -> Result<String, AuthenticationError> {
+        self.client_container
+            .read()
+            .client
+            .as_ref()
+            .ok_or(AuthenticationError::ClientMissing)
+            .map(|client| client.homeserver())
+    }
+
+    /// The OIDC Provider that is trusted by the homeserver. `nil` when
+    /// not configured.
+    pub fn authentication_issuer(&self) -> Result<Option<String>, AuthenticationError> {
+        self.client_container
+            .read()
+            .client
+            .as_ref()
+            .ok_or(AuthenticationError::ClientMissing)
+            .map(|client| client.authentication_issuer())
+    }
+
+    /// Whether the current homeserver supports the password login flow.
+    pub fn supports_password_login(&self) -> Result<bool, AuthenticationError> {
+        self.client_container
+            .read()
+            .client
+            .as_ref()
+            .ok_or(AuthenticationError::ClientMissing)
+            .and_then(|client| client.supports_password_login().map_err(AuthenticationError::from))
+    }
+
+    /// Updates the server to authenticate with the specified homeserver.
+    pub fn use_server(&self, server_name: String) -> Result<(), AuthenticationError> {
+        // Construct a username as the builder currently requires one.
+        let username = format!("@auth:{}", server_name);
+        let client = Arc::new(ClientBuilder::new())
+            .base_path(self.base_path.clone())
+            .username(username)
+            .build()
+            .map_err(AuthenticationError::from)?;
+
+        let mut client_container = self.client_container.write();
+        client_container.client = Some(client);
+        Ok(())
+    }
+
+    /// Performs a password login using the current homeserver.
+    pub fn login(
+        &self,
+        username: String,
+        password: String,
+    ) -> Result<Arc<Client>, AuthenticationError> {
+        match self.client_container.read().client.as_ref() {
+            Some(client) => {
+                let homeserver_url = client.homeserver();
+
+                // Create a new client to setup the store path for the username
+                let client = Arc::new(ClientBuilder::new())
```

Comment:
Yes that's correct, it's totally a workaround 🙈. Jonas and Ben had a conversation around whether the client store should be set at all until after login has succeeded. Amongst other things this would help if the user ID returned by the homeserver differed to what was used to log in.

This all seemed like it was a larger thought than just the bindings to me, so I opted for this for now. However if its something I could address in this PR, would be happy to be given some direction and have a go 🙂